File: py/src/braintrust/functions/stream.py
"""
This module provides classes and functions for handling Braintrust streams.

A Braintrust stream is a wrapper around a generator of `BraintrustStreamChunk`,
with utility methods to make them easy to log and convert into various formats.
"""

import dataclasses
import json
from itertools import tee
from typing import Generator, List, Literal, Union

from sseclient import SSEClient


@dataclasses.dataclass
class BraintrustTextChunk:
    """
    A chunk of text data from a Braintrust stream.
    """

    data: str
    type: Literal["text_delta"] = "text_delta"


@dataclasses.dataclass
class BraintrustJsonChunk:
    """
    A chunk of JSON data from a Braintrust stream.
    """

    data: str
    type: Literal["json_delta"] = "json_delta"


@dataclasses.dataclass
class BraintrustErrorChunk:
    """
    An error chunk from a Braintrust stream.
    """

    data: str
    type: Literal["error"] = "error"


@dataclasses.dataclass
class BraintrustConsoleChunk:
    """
    A console chunk from a Braintrust stream.
    """

    message: str
    stream: Literal["stderr", "stdout"]
    type: Literal["console"] = "console"


@dataclasses.dataclass
class BraintrustProgressChunk:
    """
    A progress chunk from a Braintrust stream.
    """

    data: str
    id: str
    object_type: str
    format: str
    output_type: str
    name: str
    event: Literal["json_delta", "text_delta"]
    type: Literal["progress"] = "progress"


class BraintrustInvokeError(ValueError):
    """
    An error that occurs during a Braintrust stream.
    """

    pass


BraintrustStreamChunk = Union[BraintrustTextChunk, BraintrustJsonChunk, BraintrustErrorChunk]


class BraintrustStream:
    """
    A Braintrust stream. This is a wrapper around a generator of `BraintrustStreamChunk`,
    with utility methods to make them easy to log and convert into various formats.
    """

    def __init__(self, base_stream: Union[SSEClient, List[BraintrustStreamChunk]]):
        """
        Initialize a BraintrustStream.

        Args:
            base_stream: Either an SSEClient or a list of BraintrustStreamChunks.
        """
        if isinstance(base_stream, SSEClient):
            self.stream = self._parse_sse_stream(base_stream)
        else:
            self.stream = base_stream
        self._memoized_final_value = None

    def _parse_sse_stream(self, sse_client: SSEClient) -> Generator[BraintrustStreamChunk, None, None]:
        """
        Parse an SSE stream into BraintrustStreamChunks.

        Args:
            sse_client: The SSEClient to parse.

        Yields:
            BraintrustStreamChunk: Parsed chunks from the SSE stream.
        """
        for event in sse_client.events():
            if event.event == "text_delta":
                yield BraintrustTextChunk(data=json.loads(event.data))
            elif event.event == "json_delta":
                yield BraintrustJsonChunk(data=event.data)
            elif event.event == "error":
                yield BraintrustErrorChunk(data=json.loads(event.data))
            elif event.event == "console":
                event_data = json.loads(event.data)
                yield BraintrustConsoleChunk(
                    message=event_data["message"],
                    stream=event_data["stream"],
                )
            elif event.event == "progress":
                event_data = json.loads(event.data)
                yield BraintrustProgressChunk(
                    data=event_data["data"],
                    id=event_data["id"],
                    object_type=event_data["object_type"],
                    format=event_data["format"],
                    output_type=event_data["output_type"],
                    name=event_data["name"],
                    event=event_data["event"],
                )

    def copy(self):
        """
        Copy the stream. This returns a new stream that shares the same underlying
        generator (via `tee`). Since generators are consumed in Python, use `copy()` if you
        need to use the stream multiple times.

        Returns:
            BraintrustStream: A new stream that you can independently consume.
        """
        current_stream = self.stream
        self.stream, new_stream = tee(current_stream)
        return BraintrustStream(new_stream)

    def final_value(self):
        """
        Get the final value of the stream. The final value is the concatenation of all
        the chunks in the stream, deserialized into a string or object, depending on
        the value's type.

        This function consumes the stream, so if you need to use the stream multiple
        times, you should call `copy()` first.

        Returns:
            The final value of the stream.
        """
        if self._memoized_final_value is None:
            self._memoized_final_value = parse_stream(self)
        return self._memoized_final_value

    def __iter__(self):
        """
        Iterate over the stream chunks.

        Yields:
            BraintrustStreamChunk: The next chunk in the stream.
        """
        yield from self.stream


def parse_stream(stream: BraintrustStream):
    """
    Parse a BraintrustStream into its final value.

    Args:
        stream: The BraintrustStream to parse.

    Returns:
        The final value of the stream.
    """
    text_chunks = []
    json_chunks = []

    for chunk in stream:
        if isinstance(chunk, BraintrustTextChunk):
            text_chunks.append(chunk.data)
        elif isinstance(chunk, BraintrustJsonChunk):
            json_chunks.append(chunk.data)
        elif isinstance(chunk, BraintrustErrorChunk):
            raise BraintrustInvokeError(chunk.data)
        elif isinstance(chunk, BraintrustProgressChunk) or isinstance(chunk, BraintrustConsoleChunk):
            pass
        else:
            raise ValueError(f"Unknown chunk type (you may need to update the SDK): {type(chunk)}")

    if json_chunks:
        return json.loads("".join(json_chunks))
    elif text_chunks:
        return "".join(text_chunks)
    else:
        return None

File: js/src/functions/stream.ts
import {
  CallEventSchema,
  callEventSchema,
  sseConsoleEventDataSchema,
  sseProgressEventDataSchema,
} from "@braintrust/core/typespecs";
import {
  createParser,
  EventSourceParser,
  ParsedEvent,
  ReconnectInterval,
} from "eventsource-parser";
import { z } from "zod";

export const braintrustStreamChunkSchema = z.union([
  z.object({
    type: z.literal("text_delta"),
    data: z.string(),
  }),
  z.object({
    type: z.literal("reasoning_delta"),
    data: z.string(),
  }),
  z.object({
    type: z.literal("json_delta"),
    data: z.string(),
  }),
  z.object({
    type: z.literal("error"),
    data: z.string(),
  }),
  z.object({
    type: z.literal("console"),
    data: sseConsoleEventDataSchema,
  }),
  z.object({
    type: z.literal("progress"),
    data: sseProgressEventDataSchema,
  }),
  z.object({
    type: z.literal("start"),
    data: z.string(),
  }),
  z.object({
    type: z.literal("done"),
    data: z.string(),
  }),
]);

/**
 * A chunk of data from a Braintrust stream. Each chunk type matches
 * an SSE event type.
 */
export type BraintrustStreamChunk = z.infer<typeof braintrustStreamChunkSchema>;

/**
 * A Braintrust stream. This is a wrapper around a ReadableStream of `BraintrustStreamChunk`,
 * with some utility methods to make them easy to log and convert into various formats.
 */
export class BraintrustStream {
  private stream: ReadableStream<BraintrustStreamChunk>;
  private memoizedFinalValue: Promise<unknown> | undefined;
  private signal: AbortSignal | undefined;

  constructor(
    baseStream: ReadableStream<Uint8Array>,
    opts?: { signal?: AbortSignal },
  );
  constructor(stream: ReadableStream<string>, opts?: { signal?: AbortSignal });
  constructor(
    stream: ReadableStream<BraintrustStreamChunk>,
    opts?: { signal?: AbortSignal },
  );
  constructor(
    baseStream:
      | ReadableStream<Uint8Array>
      | ReadableStream<string>
      | ReadableStream<BraintrustStreamChunk>,
    { signal }: { signal?: AbortSignal } = {},
  ) {
    this.signal = signal;
    this.stream = baseStream.pipeThrough(btStreamParser(), { signal });
  }

  /**
   * Copy the stream. This returns a new stream that shares the same underlying
   * stream (via `tee`). Since streams are consumed in Javascript, use `copy()` if you
   * need to use the stream multiple times.
   *
   * @returns A new stream that you can independently consume.
   */
  public copy(): BraintrustStream {
    // Once a stream is tee'd, it is essentially consumed, so we need to replace our own
    // copy of it.
    const [newStream, copyStream] = this.stream.tee();
    this.stream = copyStream;
    return new BraintrustStream(newStream, { signal: this.signal });
  }

  /**
   * Get the underlying ReadableStream.
   *
   * @returns The underlying ReadableStream<BraintrustStreamChunk>.
   */
  public toReadableStream(): ReadableStream<BraintrustStreamChunk> {
    return this.stream;
  }

  /**
   * Returns an async iterator for the BraintrustStream.
   * This allows for easy consumption of the stream using a for-await...of loop.
   *
   * @returns An async iterator that yields BraintrustStreamChunk objects.
   */
  [Symbol.asyncIterator](): AsyncIterator<BraintrustStreamChunk> {
    const reader = this.stream.getReader();
    return {
      async next() {
        const { done, value } = await reader.read();
        if (done) {
          reader.releaseLock();
          return { done: true, value: undefined };
        }
        return { done: false, value };
      },
      async return() {
        reader.releaseLock();
        return { done: true, value: undefined };
      },
      async throw(error: unknown) {
        reader.releaseLock();
        throw error;
      },
    };
  }

  /**
   * Get the final value of the stream. The final value is the concatenation of all
   * the chunks in the stream, deserialized into a string or JSON object, depending on
   * the value's type.
   *
   * This function returns a promise that resolves when the stream is closed, and
   * contains the final value. Multiple calls to `finalValue()` will return the same
   * promise, so it is safe to call this multiple times.
   *
   * This function consumes the stream, so if you need to use the stream multiple
   * times, you should call `copy()` first.
   *
   * @returns A promise that resolves with the final value of the stream or `undefined` if the stream is empty.
   */
  public finalValue(): Promise<unknown> {
    if (this.memoizedFinalValue) {
      return this.memoizedFinalValue;
    }
    this.memoizedFinalValue = new Promise((resolve, reject) => {
      this.stream
        .pipeThrough(createFinalValuePassThroughStream(resolve, reject), {
          signal: this.signal,
        })
        .pipeTo(devNullWritableStream(), { signal: this.signal })
        .catch(reject);
    });
    return this.memoizedFinalValue;
  }

  static parseRawEvent(event: CallEventSchema): BraintrustStreamChunk {
    switch (event.event) {
      case "text_delta":
        return {
          type: "text_delta",
          data: JSON.parse(event.data),
        };
      case "reasoning_delta":
        return {
          type: "reasoning_delta",
          data: JSON.parse(event.data),
        };
      case "json_delta":
        return {
          type: "json_delta",
          data: event.data,
        };
      case "error":
        return {
          type: "error",
          data: JSON.parse(event.data),
        };
      case "progress":
        return {
          type: "progress",
          data: sseProgressEventDataSchema.parse(JSON.parse(event.data)),
        };
      case "console":
        return {
          type: "console",
          data: sseConsoleEventDataSchema.parse(JSON.parse(event.data)),
        };
      case "start":
        return {
          type: "start",
          data: "",
        };
      case "done":
        return {
          type: "done",
          data: "",
        };
      default: {
        const _event: never = event;
        throw new Error(`Unknown event type ${JSON.stringify(_event)}`);
      }
    }
  }

  static serializeRawEvent(event: BraintrustStreamChunk): CallEventSchema {
    switch (event.type) {
      case "text_delta":
        return {
          event: "text_delta",
          data: JSON.stringify(event.data),
        };
      case "reasoning_delta":
        return {
          event: "reasoning_delta",
          data: JSON.stringify(event.data),
        };
      case "json_delta":
        return {
          event: "json_delta",
          data: event.data,
        };
      case "error":
        return {
          event: "error",
          data: JSON.stringify(event.data),
        };
      case "progress":
        return {
          event: "progress",
          data: JSON.stringify(event.data),
        };
      case "console":
        return {
          event: "console",
          data: JSON.stringify(event.data),
        };
      case "start":
        return {
          event: "start",
          data: "",
        };
      case "done":
        return {
          event: "done",
          data: "",
        };
      default: {
        const _event: never = event;
        throw new Error(`Unknown event type ${JSON.stringify(_event)}`);
      }
    }
  }
}

function btStreamParser() {
  const decoder = new TextDecoder();
  let parser: EventSourceParser;
  return new TransformStream<
    Uint8Array | string | BraintrustStreamChunk,
    BraintrustStreamChunk
  >({
    async start(controller) {
      parser = createParser((event: ParsedEvent | ReconnectInterval) => {
        if (event.type === "reconnect-interval") {
          return;
        }
        const parsed = callEventSchema.safeParse(event);
        if (!parsed.success) {
          throw new Error(`Failed to parse event: ${parsed.error}`);
        }
        controller.enqueue(BraintrustStream.parseRawEvent(parsed.data));
      });
    },
    async transform(chunk, controller) {
      if (chunk instanceof Uint8Array) {
        parser.feed(decoder.decode(chunk));
      } else if (typeof chunk === "string") {
        parser.feed(chunk);
      } else {
        controller.enqueue(chunk);
      }
    },
    async flush(controller) {
      controller.terminate();
    },
  });
}

/**
 * Create a stream that passes through the final value of the stream. This is
 * used to implement `BraintrustStream.finalValue()`.
 *
 * @param onFinal A function to call with the final value of the stream.
 * @returns A new stream that passes through the final value of the stream.
 */
export function createFinalValuePassThroughStream<
  T extends BraintrustStreamChunk | string | Uint8Array,
>(
  onFinal: (result: unknown) => void,
  onError: (error: unknown) => void,
): TransformStream<T, BraintrustStreamChunk> {
  const decoder = new TextDecoder();
  const textChunks: string[] = [];
  const jsonChunks: string[] = [];
  const reasoningChunks: string[] = [];
  const transformStream = new TransformStream<T, BraintrustStreamChunk>({
    transform(chunk, controller) {
      if (typeof chunk === "string") {
        textChunks.push(chunk);
        controller.enqueue({
          type: "text_delta",
          data: chunk,
        });
      } else if (chunk instanceof Uint8Array) {
        textChunks.push(decoder.decode(chunk));
        controller.enqueue({
          type: "text_delta",
          data: decoder.decode(chunk),
        });
      } else if (braintrustStreamChunkSchema.safeParse(chunk).success) {
        const chunkType = chunk.type;
        switch (chunkType) {
          case "text_delta":
            textChunks.push(chunk.data);
            break;
          case "json_delta":
            jsonChunks.push(chunk.data);
            break;
          case "reasoning_delta":
            reasoningChunks.push(chunk.data);
            break;
          case "error":
            onError(chunk.data);
            break;
          case "progress":
          case "start":
          case "done":
          case "console":
            break;
          default:
            const _type: never = chunkType;
            onError(`Unknown chunk type: ${_type}`);
        }
        controller.enqueue(chunk);
      } else {
        onError(`Unknown chunk type ${JSON.stringify(chunk)}`);
      }
    },
    flush(controller) {
      if (jsonChunks.length > 0) {
        // If we received both text and json deltas in the same stream, we
        // only return the json delta
        onFinal(JSON.parse(jsonChunks.join("")));
      } else if (textChunks.length > 0) {
        onFinal(textChunks.join(""));
      } else if (reasoningChunks.length > 0) {
        onFinal(reasoningChunks.join(""));
      } else {
        onFinal(undefined);
      }

      controller.terminate();
    },
  });

  return transformStream;
}

export function devNullWritableStream(): WritableStream {
  return new WritableStream({
    write(chunk) {},
    close() {},
    abort(reason) {},
    start(controller) {},
  });
}
